### PR TITLE
Add framework to diagnostic from compilation

### PR DESCRIPTION
**Pull request by troydai**
Addressing #1722 

Adding framework information to the diagnostic message print to the Console. It is supposed to be picked up by VS and flow to WTE with the framework information.

We used to use `CSharpDiagnosticFormatter` to format the `Diagnostic` into message. However it is basically not extensible (with one virtual method do it all). 
https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Diagnostic/DiagnosticFormatter.cs

I added `RoslynDiagnosticFormatter` to serialize the message using the same logic as `DiagnosticFormatter` but insert the framework in it.

The information represents framework is it's identity. I was suggested by @BillHiebert that version is not required. I won't be able to insert the friendly name because `Microsoft.Framework.Runtime.Roslyn.Common` doesn't depends on `Microsoft.Framework.Runtime` to avoid circular dependencies. But the framework identity should be enough to parse it back to `FrameworkName`.

Here's a sample output:
```
$ dnu build
Microsoft .NET Development Utility CLR-x86-1.0.0-t150716000045


Building playground for DNX,Version=v4.5.1
  Using Project dependency playground 1.0.0
    Source: C:\code\playground\project.json

  Using Assembly dependency framework/fx/mscorlib 4.0.0.0
    Source: C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.1\mscorlib.dll

  Using Assembly dependency framework/fx/System 4.0.0.0
    Source: C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.1\System.dll

  Using Assembly dependency framework/fx/System.Core 4.0.0.0
    Source: C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.1\System.Core.dll

  Using Assembly dependency framework/fx/Microsoft.CSharp 4.0.0.0
    Source: C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5.1\Microsoft.CSharp.dll

C:\code\playground\Program.cs(6,5): DNX error CS1519: Invalid token '}' in class, struct, or interface member declaration

Building playground for DNXCore,Version=v5.0
  Using Project dependency playground 1.0.0
    Source: C:\code\playground\project.json

  Using Package dependency System.Runtime 4.0.20-beta-23110
    Source: C:\Users\Xueyuan\.dnx\packages\System.Runtime\4.0.20-beta-23110
    File: ref\dotnet\System.Runtime.dll

  Using Package dependency System.Private.Uri 4.0.0-beta-23110
    Source: C:\Users\Xueyuan\.dnx\packages\System.Private.Uri\4.0.0-beta-23110

C:\code\playground\Program.cs(6,5): DNXCore error CS1519: Invalid token '}' in class, struct, or interface member declaration

Building playground for .NETPlatform,Version=v5.0
  Using Project dependency playground 1.0.0
    Source: C:\code\playground\project.json

  Using Package dependency System.Runtime 4.0.20-beta-23110
    Source: C:\Users\Xueyuan\.dnx\packages\System.Runtime\4.0.20-beta-23110
    File: ref\dotnet\System.Runtime.dll

C:\code\playground\Program.cs(6,5): .NETPlatform error CS1519: Invalid token '}' in class, struct, or interface member declaration
``` 

Error message from runtime error: 
```
System.IO.FileLoadException: Could not load file or assembly 'ConsoleApp1' or one of its dependencies. General Exception (Exception from HRESULT: 0x80131500)
File name: 'ConsoleApp1' ---> Microsoft.Framework.Runtime.Roslyn.RoslynCompilationException: C:\code\playground\ConsoleApp1\src\ConsoleApp1\Program.cs(12,7): DNX,Version=v4.5.1 error CS1001: Identifier expected
C:\code\playground\ConsoleApp1\src\ConsoleApp1\Program.cs(12,7): DNX,Version=v4.5.1 error CS1002: ; expected
C:\code\playground\ConsoleApp1\src\ConsoleApp1\Program.cs(12,6): DNX,Version=v4.5.1 error CS0246: The type or namespace name 'a' could not be found (are you missing a using directive or an assembly reference?)
   at Microsoft.Framework.Runtime.Roslyn.RoslynProjectReference.Load(IAssemblyLoadContext loadContext) in C:\code\DNX\src\Microsoft.Framework.Runtime.Roslyn\RoslynProjectReference.cs:line 117
   at Microsoft.Framework.Runtime.Loader.ProjectAssemblyLoader.Load(AssemblyName assemblyName, IAssemblyLoadContext loadContext) in C:\code\DNX\src\Microsoft.Framework.Runtime\Loader\ProjectAssemblyLoader.cs:line 66
   at Microsoft.Framework.Runtime.Loader.ProjectAssemblyLoader.Load(AssemblyName assemblyName) in C:\code\DNX\src\Microsoft.Framework.Runtime\Loader\ProjectAssemblyLoader.cs:line 28
   at dnx.host.LoaderContainer.Load(AssemblyName assemblyName) in C:\code\DNX\src\dnx.host\LoaderContainer.cs:line 38
   at dnx.host.DefaultLoadContext.LoadAssembly(AssemblyName assemblyName) in C:\code\DNX\src\dnx.host\DefaultLoadContext.cs:line 20
   at Microsoft.Framework.Runtime.Loader.AssemblyLoaderCache.GetOrAdd(AssemblyName name, Func`2 factory) in C:\code\DNX\src\Microsoft.Framework.Runtime.Loader\AssemblyLoaderCache.cs:line 37
   at Microsoft.Framework.Runtime.Loader.LoadContext.LoadAssemblyImpl(AssemblyName assemblyName) in C:\code\DNX\src\Microsoft.Framework.Runtime.Loader\LoadContext.cs:line 144
   at Microsoft.Framework.Runtime.Loader.LoadContext.ResolveAssembly(Object sender, ResolveEventArgs args) in C:\code\DNX\src\Microsoft.Framework.Runtime.Loader\LoadContext.cs:line 236
   at System.AppDomain.OnAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIn
trospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forInt
rospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean fo
rIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.Assembly.Load(AssemblyName assemblyRef)
   at Microsoft.Framework.Runtime.Loader.LoadContext.Load(AssemblyName assemblyName) in C:\code\DNX\src\Microsoft.Framework.Runtime.Loader\LoadContext.cs:line 133
   at Microsoft.Framework.Runtime.AssemblyLoadContextExtensions.Load(IAssemblyLoadContext loadContext, String name) in C:\code\DNX\src\Microsoft.Framework.Runtime.Abstractions\AssemblyLoadContextExtensions.cs:line 9
   at Microsoft.Framework.Runtime.DefaultHost.GetEntryPoint(String applicationName) in C:\code\DNX\src\Microsoft.Framework.Runtime\DefaultHost.cs:line 95
   at Microsoft.Framework.ApplicationHost.Program.ExecuteMain(DefaultHost host, String applicationName, String[] args) in C:\code\DNX\src\Microsoft.Framework.ApplicationHost\Program.cs:line 210
   at Microsoft.Framework.ApplicationHost.Program.Main(String[] args) in C:\code\DNX\src\Microsoft.Framework.ApplicationHost\Program.cs:line 90
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Framework.Runtime.Common.EntryPointExecutor.Execute(Assembly assembly, String[] args, IServiceProvider serviceProvider) in C:\code\DNX\src\Microsoft.Framework.Runtime.Sources\Impl\EntryPointExecutor.cs:line 43
   at dnx.host.Bootstrapper.RunAsync(List`1 args, IRuntimeEnvironment env, FrameworkName targetFramework) in C:\code\DNX\src\dnx.host\Bootstrapper.cs:line 100
   at dnx.host.RuntimeBootstrapper.ExecuteAsync(String[] args, FrameworkName targetFramework) in C:\code\DNX\src\dnx.host\RuntimeBootstrapper.cs:line 144
   at dnx.host.RuntimeBootstrapper.Execute(String[] args, FrameworkName targetFramework) in C:\code\DNX\src\dnx.host\RuntimeBootstrapper.cs:line 26


Microsoft.Framework.Runtime.Roslyn.RoslynCompilationException: C:\code\playground\ConsoleApp1\src\ConsoleApp1\Program.cs(12,7): DNX,Version=v4.5.1 error CS1001: Identifier expected
C:\code\playground\ConsoleApp1\src\ConsoleApp1\Program.cs(12,7): DNX,Version=v4.5.1 error CS1002: ; expected
C:\code\playground\ConsoleApp1\src\ConsoleApp1\Program.cs(12,6): DNX,Version=v4.5.1 error CS0246: The type or namespace name 'a' could not be found (are you missing a using directive or an assembly reference?)
   at Microsoft.Framework.Runtime.Roslyn.RoslynProjectReference.Load(IAssemblyLoadContext loadContext) in C:\code\DNX\src\Microsoft.Framework.Runtime.Roslyn\RoslynProjectReference.cs:line 117
   at Microsoft.Framework.Runtime.Loader.ProjectAssemblyLoader.Load(AssemblyName assemblyName, IAssemblyLoadContext loadContext) in C:\code\DNX\src\Microsoft.Framework.Runtime\Loader\ProjectAssemblyLoader.cs:line 66
   at Microsoft.Framework.Runtime.Loader.ProjectAssemblyLoader.Load(AssemblyName assemblyName) in C:\code\DNX\src\Microsoft.Framework.Runtime\Loader\ProjectAssemblyLoader.cs:line 28
   at dnx.host.LoaderContainer.Load(AssemblyName assemblyName) in C:\code\DNX\src\dnx.host\LoaderContainer.cs:line 38
   at dnx.host.DefaultLoadContext.LoadAssembly(AssemblyName assemblyName) in C:\code\DNX\src\dnx.host\DefaultLoadContext.cs:line 20
   at Microsoft.Framework.Runtime.Loader.AssemblyLoaderCache.GetOrAdd(AssemblyName name, Func`2 factory) in C:\code\DNX\src\Microsoft.Framework.Runtime.Loader\AssemblyLoaderCache.cs:line 37
   at Microsoft.Framework.Runtime.Loader.LoadContext.LoadAssemblyImpl(AssemblyName assemblyName) in C:\code\DNX\src\Microsoft.Framework.Runtime.Loader\LoadContext.cs:line 144
   at Microsoft.Framework.Runtime.Loader.LoadContext.ResolveAssembly(Object sender, ResolveEventArgs args) in C:\code\DNX\src\Microsoft.Framework.Runtime.Loader\LoadContext.cs:line 236
   at System.AppDomain.OnAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)
```

### CODE REVIEW

File: src/Microsoft.Framework.Runtime.Roslyn.Common/RoslynDiagnosticFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                    else
+                    {
+                        path = span.Path;
+                        basePath = null;
+                    }
+
+                    var start = mappedSpan.Span.Start;
+                    var framework = targetFramework == null ? string.Empty : $" {targetFramework.Identifier}";
+                    return $"{path}({start.Line + 1},{start.Character + 1}):{framework} {GetMessagePrefix(diagnostic)}: {diagnostic.GetMessage(culture)}";
+
```
Comment (BrennanConroy) [inline src/Microsoft.Framework.Runtime.Roslyn.Common/RoslynDiagnosticFormatter.cs]: remove newline

**Comment by troydai** [comment]
@BillHiebert says it works for him.

Ping @davidfowl

**Comment by troydai** [comment]
Need sign off @davidfowl @muratg @BrennanConroy @victorhurdugaci @anurse 
Verified by @BillHiebert

**Comment by BrennanConroy** [comment]
Could you remove the goto?

Sent from my HTC

----- Reply message -----

**Comment by davidfowl** [comment]
This change seems to affect everything doing compilation. Did you test:
- runtime compilation
- DTH compilation 
- dnu build 
- razor build
- Helios error page
- diagnostics error page

**Comment by anurse** [comment]
Looks good to me, I have similar concerns as to verifying it in all contexts. Most importantly, does this play well in the VS Output Window and Error List? The Output Window in VS has a regex it uses to sniff out things to pull into the Error List and we should make sure that if these messages are going there, we are behaving correctly.

**Comment by davidfowl** [comment]
So this requires a tooling change? I wonder why? What happens in VS RC?

**Comment by troydai** [comment]
I'll go through these error handling again tonight. Before that I already verified a few things before sending the PR:
- DTH compilation 
Output window and Error list both looks fine.
- Dnu build 
As I showed in the description, the difference is the compilation error message contains framework full names in it now.

**Comment by troydai** [comment]
In terms of the tooling change, @BillHiebert can give a better explanation.

**Comment by BillHiebert** [comment]
Previous tooling (RC and before my change), looked for the friendly framework name when matching the error to a language service project. RC will continue to work as before and route the error to the error list rather than combine it will intellisense errors. My tooling change was to expand the framework match to also check for a matching moniker.

**Comment by anurse** [comment]
This reminds me that we should try, as much as possible, to move all of our framework parsing code to use [NuGet.Frameworks](https://github.com/NuGet/NuGet3/tree/dev/src/NuGet.Frameworks). Perhaps using source-inclusion. We'll have to come up with a convenient way to distribute it (and probably NuGet.Versioning for parsing SemVers as well) so that code can use it without interfering with anyone else who wants to use it.
&lt;/off-topic>

**Comment by troydai** [comment]
Error List looks fine:

image

**Comment by troydai** [comment]
Output Window

image

**Comment by troydai** [comment]
Output Window:

image

**Comment by davidfowl** [comment]
@BillHiebert Why are 2 messages showing up in the error list? Isn't that what this entire fix is about?

**Comment by troydai** [comment]
Compilation error shown in diagnostic page - framework name shows up:

image


Razor compilation error - the framework name doesn't show up:

image

**Comment by BillHiebert** [comment]
@davidfowl  - the change ensures that VS knows which framework the error is associated with and we can push it to the correct language service which will de-dupe it with existing intellisense errors.

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**